File: HexaFlipTests/Source/AI/JCSGameAlgorithmTest.m
//
//  JCSFlipAlgoTest.m
//  HexaFlip
//
//  Created by Christian Schuster on 20.07.12.
//  Copyright (c) 2012 Christian Schuster. All rights reserved.
//

#import "JCSGameAlgorithm.h"
#import "JCSNegamaxGameAlgorithm.h"
#import "JCSNegaScoutGameAlgorithm.h"
#import "JCSRandomGameAlgorithm.h"
#import "JCSFlipCellState.h"
#import "JCSFlipGameState.h"
#import "JCSFlipMove.h"
#import "JCSFlipGameStatePSRHeuristic.h"
#import "JCSFlipGameStatePossessionHeuristic.h"

@interface JCSFlipAlgoTest : XCTestCase
@end

@implementation JCSFlipAlgoTest

- (void)testNegamax2VsNegamax1 {
    id<JCSGameHeuristic> paranoid = [[JCSFlipGameStatePSRHeuristic alloc] initWithPossession:1 safety:3 randomness:0];
    id<JCSGameAlgorithm> algoA = [[JCSNegamaxGameAlgorithm alloc] initWithDepth:2 heuristic:paranoid];
    id<JCSGameAlgorithm> algoB = [[JCSNegamaxGameAlgorithm alloc] initWithDepth:1 heuristic:paranoid];
    NSInteger size = 4;
    [self testAlgorithm:algoA againstAlgorithm:algoB withBoardSize:size];
}


- (void)testNegamax3VsNegamax2 {
    id<JCSGameHeuristic> careless = [[JCSFlipGameStatePSRHeuristic alloc] initWithPossession:1 safety:0 randomness:2];
    id<JCSGameHeuristic> paranoid = [[JCSFlipGameStatePSRHeuristic alloc] initWithPossession:1 safety:3 randomness:0];
    id<JCSGameAlgorithm> algoA = [[JCSNegamaxGameAlgorithm alloc] initWithDepth:3 heuristic:careless];
    id<JCSGameAlgorithm> algoB = [[JCSNegamaxGameAlgorithm alloc] initWithDepth:2 heuristic:paranoid];
    NSInteger size = 4;
    [self testAlgorithm:algoA againstAlgorithm:algoB withBoardSize:size];
}

- (void)testNegamax3VsRandom {
    id<JCSGameHeuristic> safe = [[JCSFlipGameStatePSRHeuristic alloc] initWithPossession:1 safety:0.5 randomness:0.25];
    id<JCSGameAlgorithm> algoA = [[JCSNegamaxGameAlgorithm alloc] initWithDepth:3 heuristic:safe];
    id<JCSGameAlgorithm> algoB = [[JCSRandomGameAlgorithm alloc] initWithSeed:9391829];
    NSInteger size = 4;
    [self testAlgorithm:algoA againstAlgorithm:algoB withBoardSize:size];
}

- (void)testNegaScoutVsNegaScout {
    id<JCSGameHeuristic> possessive = [[JCSFlipGameStatePossessionHeuristic alloc] init];
    id<JCSGameAlgorithm> algoA = [[JCSNegaScoutGameAlgorithm alloc] initWithDepth:4 heuristic:possessive];
    id<JCSGameAlgorithm> algoB = [[JCSNegaScoutGameAlgorithm alloc] initWithDepth:4 heuristic:possessive];
    NSInteger size = 4;
    [self testAlgorithm:algoA againstAlgorithm:algoB withBoardSize:size];
}

- (void)testAlgorithm:(id<JCSGameAlgorithm>)algoA againstAlgorithm:(id<JCSGameAlgorithm>)algoB withBoardSize:(NSInteger)size {
    JCSFlipGameState *state = [[JCSFlipGameState alloc] initWithSize:size playerToMove:JCSFlipPlayerSideA];
    
    while (!state.leaf) {
        id<JCSGameAlgorithm> algo = (state.playerToMove == JCSFlipPlayerSideA ? algoA : algoB);
        id<JCSMove> move = [algo moveAtNode:state];
        XCTAssertNotNil(move, @"move returned by algorithm must not be nil for non-leaf node");
        [state pushMove:move];
        NSLog(@"\n%@", state);
    }
    NSLog(@"done, final scores %d:%d", state.cellCountPlayerA, state.cellCountPlayerB);
    
    XCTAssertTrue(state.status == JCSFlipGameStatusPlayerAWon || state.status == JCSFlipGameStatusPlayerBWon || state.status == JCSFlipGameStatusDraw);
}

@end

File: HexaFlipTests/Source/AI/JCSGameAlgorithmTest.swift
//
//  JCSGameAlgorithmTest.swift
//  HexaFlip
//
//  Created by Christian Schuster on 27.09.14.
//  Copyright (c) 2014 Christian Schuster. All rights reserved.
//

import Foundation
import XCTest


class JCSGameAlgorithmTest : XCTestCase {
    
    func testNegamax2VsNegamax1() {
        let paranoid = JCSFlipGameStatePSRHeuristic(possession: 1, safety: 3, randomness: 0)
        let algoA = JCSNegamaxGameAlgorithm(depth: 2, heuristic: paranoid)
        let algoB = JCSNegamaxGameAlgorithm(depth: 1, heuristic: paranoid)
        testAlgorithm(algoA, algoB, boardSize: 4)
    }
    
    func testNegamax3VsNegamax2() {
        let careless = JCSFlipGameStatePSRHeuristic(possession: 1, safety: 0, randomness: 2)
        let paranoid = JCSFlipGameStatePSRHeuristic(possession: 1, safety: 3, randomness: 0)
        let algoA = JCSNegamaxGameAlgorithm(depth: 3, heuristic: careless)
        let algoB = JCSNegamaxGameAlgorithm(depth: 2, heuristic: paranoid)
        testAlgorithm(algoA, algoB, boardSize: 4)
    }
    
    func testNegamax3VsRandom() {
        let safe = JCSFlipGameStatePSRHeuristic(possession: 1, safety: 0.5, randomness: 0.25)
        let algoA = JCSNegamaxGameAlgorithm(depth: 3, heuristic: safe)
        let algoB = JCSRandomGameAlgorithm(seed: 9391829)
        testAlgorithm(algoA, algoB, boardSize: 4)
    }
    
    func testNegaScoutVsNegaScout() {
        let possessive = JCSFlipGameStatePossessionHeuristic()
        let algoAB = JCSNegaScoutGameAlgorithm(depth: 5, heuristic: possessive)
        testAlgorithm(algoAB, algoAB, boardSize: 4)
    }
    
    func testAlgorithm(algoA: JCSGameAlgorithm, _ algoB: JCSGameAlgorithm, boardSize: Int) {
        let state = JCSFlipGameState(size: boardSize, playerToMove: JCSFlipPlayerSide.A)
        while !state.leaf {
            let algo = (state.playerToMove == JCSFlipPlayerSide.A ? algoA : algoB)
            let move: AnyObject! = algo.moveAtNode(state)
            XCTAssertNotNil(move, "move returned by algorithm must not be nil for non-leaf node")
            state.pushMove(move)
            NSLog("\n%@", state)
        }
        NSLog("done, final scores %d:%d", state.cellCountPlayerA, state.cellCountPlayerB)
        XCTAssertTrue(state.status == JCSFlipGameStatus.PlayerAWon || state.status == JCSFlipGameStatus.PlayerBWon || state.status == JCSFlipGameStatus.Draw)
    }
}